### PR TITLE
create Forget password func and change password func

### PR DESCRIPTION
**Pull request by abhishekanalytics**

### CODE REVIEW

File: tutorial/custom_user/serializers.py [hunk 1]
```diff
@@ -10,4 +10,12 @@ class Meta:
...
         class Meta:
             model = CustomUser
-            fields = ['first_name','last_name','phone_no']
\ No newline at end of file
+            fields = ['first_name','last_name','phone_no']
+
+class ChangePasswordSerializer(serializers.Serializer):
+    old_password = serializers.CharField(required=True)
+    new_password = serializers.CharField(required=True)
+
```
Comment (pythexcel) [inline tutorial/custom_user/serializers.py:18]: Here should be a confirm password field

File: tutorial/custom_user/views.py [hunk 2]
```diff
@@ -44,4 +66,46 @@ def patch(self, request, pk, format=None):
...
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+
+class PasswordResetAPIView(APIView):
+    def post(self, request, *args, **kwargs):
+        serializer = ForgotPasswordSerializer(data=request.data)
+        if serializer.is_valid():
+            email = serializer.validated_data['email']
+            try:
+                user = CustomUser.objects.get(email=email, is_active=True)  # Use your custom user model here
```
Comment (pythexcel) [inline tutorial/custom_user/views.py:91]: Can use get_object_or_404 here

File: tutorial/custom_user/views.py [hunk 3]
```diff
@@ -44,4 +66,46 @@ def patch(self, request, pk, format=None):
...
+            except CustomUser.DoesNotExist:
+                return Response({'error': 'Invalid email address'}, status=status.HTTP_400_BAD_REQUEST)
+        
+            uid = urlsafe_base64_encode(force_bytes(user.pk))
```
Comment (pythexcel) [inline tutorial/custom_user/views.py:95]: Move the encoding and url generation logic in a separate function.
Do remember your views are just for handling request and response, Try to separate concerns from views as much you can.

File: tutorial/tutorial/settings.py [hunk 4]
```diff
@@ -144,4 +148,11 @@
     'SLIDING_TOKEN_LIFETIME': timedelta(days=30),
     'SLIDING_TOKEN_REFRESH_LIFETIME_LATE_USER': timedelta(days=1),
     'SLIDING_TOKEN_LIFETIME_LATE_USER': timedelta(days=30),
-}
\ No newline at end of file
+}
+EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
```
Comment (pythexcel) [inline tutorial/tutorial/settings.py:152]: These creds are sensitive data, Should be placed in .env

File: tutorial/custom_user/views.py [hunk 5]
```diff
@@ -75,37 +81,68 @@ def post(self, request, *args, **kwargs):
...
             user = request.user
             if user.check_password(serializer.data.get('old_password')):
-                user.set_password(serializer.data.get('new_password'))
-                user.save()
-                return Response({'message': 'Password changed successfully.'}, status=status.HTTP_200_OK)
-            return Response({'error': 'Incorrect old password.'}, status=status.HTTP_400_BAD_REQUEST)
-        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+                new_password = serializer.data.get('new_password')
+                confirm_password = serializer.data.get('confirm_password')
+                if new_password == confirm_password:
```
Comment (pythexcel) [inline tutorial/custom_user/views.py:86]: You are already validating it in the serializer, No need to check again if passwords are equal or not.

File: tutorial/custom_user/views.py [hunk 6]
```diff
@@ -75,37 +81,68 @@ def post(self, request, *args, **kwargs):
...
+
+class PasswordResetConfirmView(TemplateView):
+    template_name = 'password_reset_confirm.html'
+
+    def post(self, request, uidb64, token):
+        User = get_user_model()
+        uid = urlsafe_base64_decode(uidb64).decode()
+        user = get_object_or_404(User, pk=uid) 
+
+        if user is not None and default_token_generator.check_token(user, token):
```
Comment (pythexcel) [inline tutorial/custom_user/views.py:134]: Fix this password update logic

File: tutorial/custom_user/models.py [hunk 7]
```diff
@@ -2,6 +2,9 @@
 from django.db import models
 from enum import Enum
 from .manager import CustomUserManager
+from django.db import models
```
Comment (nodeexcel) [inline tutorial/custom_user/models.py:5]: remove duplicate import

File: tutorial/custom_user/views.py [hunk 8]
```diff
@@ -44,4 +73,80 @@ def patch(self, request, pk, format=None):
...
+                error_message = "Password reset has already been done for this user"
+            else:
+                user.set_password(new_password)
+                user.password_reset_done = True 
+                user.save()
+                return render(request, self.template_name, {'uidb64': uidb64, 'token': token})
+            return HttpResponse(error_message, status=400)
+        return HttpResponseBadRequest('You have already update password')
+
+def is_valid_password(password):
```
Comment (nodeexcel) [inline tutorial/custom_user/views.py:150]: move in utils file.

**Pull request merged**